### PR TITLE
Add more complex models

### PR DESCRIPTION
**Pull request by LindsayYoung**
Still working on this. So far, the single answer multi select fields are easy. I need to make modification for the forms.py and models for multi select fields.

closes https://github.com/18F/crt-portal/issues/104

### CODE REVIEW

**Comment by LindsayYoung** [comment]
selecting more that one option with checkboxes- is going, but I can't seem to get rid of the "---" option

So far I have tried:
 - adding `default=None` to the model elements, but that only works for a required field or a many to many field
- creating an init function in forms.py defining the fields as having `self.fields['fieldname'].empty_label = None`
- adding `fieldset.inlineLabels ul {list-style:none}` to the custom styles

**Comment by LindsayYoung** [comment]
You can run the tests with: 

    docker-compose run web python /code/crt_portal/manage.py test cts_forms

**Comment by LindsayYoung** [comment]
Hi @alexsoble this PR doesn't make style changes. Feel free to rebase locally if that is what you want. 

In the issue, I linked to a very detailed yml file: https://github.com/usdoj/crt-portal/blob/396e2a80de0ac8ffeebb32210f6912ca8c6a651a/config/default/webform.webform.benevolent_hawk.yml Please compare how this form is working to how it is described in the yml file. That will make for a better comparison than the original screen shots because the rendered form can change depending on the answers to previous questions. 

For this PR review please make sure the tests run for you and the structures make sense.

The visual changes will be:

(This assumes you have added some valad test data to http://0.0.0.0:8000/admin/cts_forms/protectedclass/add/)
Screen Shot 2019-08-20 at 1 13 10 PM

and no "----" option
Screen Shot 2019-08-20 at 1 17 58 PM

**Comment by alexsoble** [comment]
@LindsayYoung Are we most interested in the `elements` section of the YML file? That seems to have the most content; most of the other keys are set to `false`, `{}`, `""`, etc.

**Comment by LindsayYoung** [comment]
Yes, the Drupal specific stuff doesn't get ported over here is the YML in that file that describes the form:
```
report_a_violation:
    '#type': webform_section
    '#title': 'Report a violation'
    '#description': 'If you believe your (or someone else&rsquo;s) civil rights have been violated, please use the form below to contact the Department of Justice.'
  primary_complaint:
    '#type': webform_radios_other
    '#title': 'What is your primary reason for contacting the Civil Rights Division?'
    '#description': 'Please pick the reason that best fits the situation. We realize that multiple situations may apply and you&rsquo;ll be given the opportunity to explain in more detail later.'
    '#description_display': before
    '#options':
      'Denied access or removed from a location (including segregation)': 'Denied access or removed from a location (including segregation)'
      'Prevented from using a service or service terminated': 'Prevented from using a service or service terminated'
      'Denied housing or subjected to harmful living conditions': 'Denied housing or subjected to harmful living conditions'
      'Fired, not hired, demoted': 'Fired, not hired, demoted, or asked to show more documentation than required'
      'Retaliated against or otherwise mistreated for reporting an issue': 'Retaliated against or otherwise mistreated for reporting an issue'
      'Harassed, threatened, assaulted, or otherwise made to feel unsafe (including sexual harassment or assault)': 'Harassed, threatened, assaulted, or otherwise made to feel unsafe (including sexual harassment or assault)'
      'Ability to vote was impacted': 'Ability to vote was impacted'
      'Otherwise discriminated against': 'Otherwise discriminated against'
    '#other__option_label': Other
    '#other__title': 'Please describe'
    '#required': true
  protected_class:
    '#type': webform_checkboxes_other
    '#title': 'Why do you believe the discrimination took place? '
    '#description': 'Civil rights laws protect people from discrimination based on the below protected classes. Please choose which you think best describes the reason(s) for the discrimination.'
    '#description_display': before
    '#options':
      Race/color: Race/color
      'National origin (including ancestry, ethnicity, and language)': 'National origin (including ancestry, ethnicity, and language)'
      'Immigration/citizenship status (you do not have to divulge your status)': 'Immigration/citizenship status (you do not have to divulge your status)'
      Religion: Religion
      'Disability (including temporary)': 'Disability (including temporary)'
      'Sex or gender idenity (including pregnancy)': 'Sex or gender idenity (including pregnancy)'
      'Sexual orientation': 'Sexual orientation'
      'Familial, marital, or parental status': 'Familial, marital, or parental status'
      Age: Age
      'Military status': 'Military status'
      'Genetic information': 'Genetic information'
    '#other__option_label': Other
    '#other__placeholder': 'Please describe'
    '#required': true
where:
  '#type': webform_wizard_page
  '#title': Where
  '#prev_button_label': Back
  '#next_button_label': Next
  where_did_this_take_place_:
    '#type': webform_radios_other
    '#title': 'Where did this take place?'
    '#description': 'Providing the location helps us direct your issue to the most relevant part of Civil Rights Division'
    '#description_display': before
    '#options':
      'Home, potential home, or services to help with purchasing a home (banks, lenders, or other financial services)': 'Home, potential home, or services to help with purchasing a home (banks, lenders, or other financial services)'
      'Workplace or potential workplace': 'Workplace or potential workplace'
      'Educational institution (school, university), education program or educational activity (after school program or workshop)': 'Educational institution (school, university), education program or educational activity (after school program or workshop)'
      'Place of worship': 'Place of worship'
      'Retail/commercial building (store, restaurant, hotel, nightclub, theater, gym, or other commercial space)': 'Retail/commercial building (store, restaurant, hotel, nightclub, theater, gym, or other commercial space)'
      'Outdoor public space (including car, street, sidewalk, park)': 'Outdoor public space (including car, street, sidewalk, park)'
      'Voting location or ballot (including mail-in ballots)': 'Voting location or ballot (including mail-in ballots)'
      'Healthcare facility (including mental health or long-term care)': 'Healthcare facility (including mental health or long-term care)'
      'Prison, jail, or juvenile corrections facility, or while otherwise incarcerated': 'Prison, jail, or juvenile corrections facility, or while otherwise incarcerated'
      ' Another government building (courthouse, DMV, post office)': ' Another government building (courthouse, DMV, post office)'
    '#other__option_label': Other
    '#other__title': 'Please describe'
    '#required': true
  was_this_a_public_or_private_employer_:
    '#type': radios
    '#title': 'Was this a public or private employer?'
    '#options':
      'Public employer': 'Public employer -- Funded by the government like a post office, fire department, courthouse, DMV, or public school. This could be at the local, state, or federal level.'
      'Private employer': 'Private employer -- Businesses or non-profits not funded by the government such as retail stores, banks, or restaurants.'
      'I''m not sure': 'I''m not sure'
    '#states':
      visible:
        ':input[name=\"where_did_this_take_place_[radios]\"]':
          value: 'Workplace or potential workplace'
    '#states_clear': false
    '#required': true
  location_details:
    '#type': webform_contact
    '#title': 'Location or entity details (optional)'
    '#required': true
    '#states_clear': false
    '#name__access': false
    '#name__title': 'Location, organization, company, or individual name'
    '#company__access': false
    '#email__access': false
    '#phone__access': false
    '#address__access': false
    '#address_2__access': false
    '#country__access': false
  employer_size:
    '#type': radios
    '#title': 'About how large is the employer?'
    '#options':
      'Fewer than 15 employees': 'Fewer than 15 employees'
      '15 or more employees': '15 or more employees'
    '#states':
      visible:
        ':input[name=\"where_did_this_take_place_[radios]\"]':
          value: 'Workplace or potential workplace'
    '#states_clear': false
    '#required': true
  public_or_private_edu:
    '#type': radios
    '#title': 'Was this a public or private educational institution or program?'
    '#options':
      Public: 'Public -- Schools or programs funded by local, state, or the federal government'
      Private: 'Private -- Schools or programs funded privately such as charter schools, magnet schools, or faith-based colleges'
      'I''m not sure': 'I''m not sure'
    '#states':
      visible:
        ':input[name=\"where_did_this_take_place_[radios]\"]':
          value: 'Educational institution (school, university), education program or educational activity (after school program or workshop)'
    '#states_clear': false
    '#required': true
  public_or_private_facility:
    '#type': radios
    '#title': 'What kind of facility?'
    '#options':
      'State/local facility': 'State/local facility'
      'Federal facility': 'Federal facility'
      'Private facility': 'Private facility'
      'Not sure': 'Not sure'
    '#states':
      visible:
        ':input[name=\"where_did_this_take_place_[radios]\"]':
          value: 'Prison, jail, or juvenile corrections facility, or while otherwise incarcerated'
    '#states_clear': false
    '#required': true
  public_or_private_healthcare:
    '#type': radios
    '#title': 'What kind of healthcare facility?'
    '#options':
      'State/local facility': 'State/local facility'
      'Federal facility': 'Federal facility'
      'Private facility': 'Private facility'
      'Not sure': 'Not sure'
    '#states':
      visible:
        ':input[name=\"where_did_this_take_place_[radios]\"]':
          value: 'Healthcare facility (including mental health or long-term care)'
    '#states_clear': false
    '#required': true
who:
  '#type': webform_wizard_page
  '#title': Who
  '#prev_button_label': Back
  '#next_button_label': Next
  respondent_type:
    '#type': webform_radios_other
    '#title': 'Who are you reporting this violation against? (optional)'
    '#description': 'Letting us know the person or organization you&rsquo;re reporting against also helps us.'
    '#description_display': before
    '#options':
      'Employer or potential employer': 'Employer or potential employer'
      'Landlord, leasing office, or home/rental provider': 'Landlord, leasing office, or home/rental provider'
      'Police, prison guard, or other corrections staff': 'Police, prison guard, or other corrections staff'
      'Other public official (judge, voting official, or other government official)': 'Other public official (judge, voting official, or other government official)'
      'Individual(s) from school or educational program (teacher, administrator, staff, or students)': 'Individual(s) from school or educational program (teacher, administrator, staff, or students)'
      'Healthcare provider or staff': 'Healthcare provider or staff'
      'Bank or loaning service': 'Bank or loaning service'
    '#other__option_label': 'Other individual or group (family member, neighbor, stranger, or organized group)'
    '#other__title': 'Please describe (optional)'
  respondent_contact_ask:
    '#type': checkbox
    '#title': 'I’d like to include their physical address, email, or phone information '
  respondent_contact:
    '#type': webform_contact
    '#title': 'Location or entity contact details (optional)'
    '#states':
      visible:
        ':input[name=\"i_d_like_to_include_their_physical_address_email_or_phone_inform\"]':
          checked: true

more_details:
  '#type': webform_wizard_page
  '#title': 'More details'
  '#prev_button_label': Back
  '#next_button_label': Next
  violation_summary:
    '#type': textarea
    '#title': 'Please describe what happened in your own words, including any additional details you want us to know'
    '#description': 'Additional details:'
    '#description_display': before
    '#counter_type': word
    '#counter_minimum': '1'
    '#counter_maximum': '500'
    '#required': true
  when_did_this_incident_last_take_place_:
    '#type': radios
    '#title': 'When did this incident last take place?'
    '#options':
      'Within the last 6 months': 'Within the last 6 months'
      'Within the last 3 years': 'Within the last 3 years'
      'More than 3 years ago': 'More than 3 years ago'
    '#required': true
  do_you_have_reason_to_believe_others_are_experiencing_this_issue:
    '#type': webform_radios_other
    '#title': 'Do you have reason to believe others are experiencing this issue? (optional)'
    '#options':
      'No': 'No'
      'I''m not sure': 'I''m not sure'
    '#other__option_label': 'Yes'
    '#other__placeholder': 'Please describe'
contact_optional_:
  '#type': webform_wizard_page
  '#title': 'Contact (optional)'
  '#prev_button_label': Back
  '#next_button_label': Review
  how_should_we_contact_you_:
    '#type': webform_section
    '#title': 'How should we contact you?'
    '#description': |
      In order for us to follow up&nbsp; with questions or communicate the status of your report, we&rsquo;ll need to know the best person to contact.&nbsp;<br />
      You may also submit this report anonymously.&nbsp;<br />
      <br />
      We read everything we receive and appreciate learning about issues of public concern. Thank you for taking the time to report this violation.

  who_are_you_reporting_for_:
    '#type': radios
    '#title': \"Who are you reporting for?\\t\"
    '#options':
      \"I'm reporting for myself\\t\": \"I'm reporting for myself\\t\"
      'I''m reporting on behalf of another person': 'I''m reporting on behalf of another person'
      \"I prefer not to disclose\\t\": \"I prefer not to disclose\\t\"
    '#required': true
  what_is_your_relationship_to_the_person_s_affected_:
    '#type': webform_radios_other
    '#title': 'What is your relationship to the person(s) affected?'
    '#options':
      'I''m a parent or guardian of the person affected': 'I''m a parent or guardian of the person affected'
      'I''m providing professional assistance': 'I''m providing professional assistance'
      'I''m a witness to the situation': 'I''m a witness to the situation'
    '#other__option_label': Other
    '#other__title': 'Please describe'
    '#other__placeholder': 'Enter other relationship'
    '#states':
      visible:
        ':input[name=\"who_are_you_reporting_for_\"]':
          value: 'I''m reporting on behalf of another person'
    '#states_clear': false
    '#required': true
  who_should_we_contact_about_this_report_:
    '#type': webform_section
    '#title': 'Who should we contact about this report?'
    i_don_t_want_to_be_contacted:
      '#type': checkbox
      '#title': 'I don''t want to be contacted'
      '#states_clear': false
    contact_details:
      '#type': webform_contact
      '#title': 'Contact details'
      '#states':
        visible:
          ':input[name=\"i_don_t_want_to_be_contacted\"]':
            unchecked: true
      '#states_clear': false
      '#name__title': 'First name'
      '#company__title': 'Last name'
  alleged_victim:
    '#type': webform_section
    '#title': 'Person(s) who was discriminated against'
    same_information_as_above:
      '#type': checkbox
      '#title': 'Same information as above'
    alleged_victim_contact:
      '#type': webform_contact
      '#title': 'Contact information for the person who was discriminated against'
      '#states':
        invisible:
          ':input[name=\"same_information_as_above\"]':
            checked: true
      '#states_clear': false
      '#name__title': 'First name'
      '#company__title': 'Last name'
```

**Comment by alexsoble** [comment]
@LindsayYoung Thanks! I will use this YAML as a starting point for review. The YAML you posted is a bit easier to work with because the form data in the link was all compressed into a single line, was hard to read. 

I know @uchmok has a background leading QA teams ... I wonder if she can help us find a more user-friendly way to express what the expected form behavior should look like. 

One thing I'm noticing as I review the form is that some of the questions marked with `required: true` in the YML are not being enforced as required in the form. For example, I am able to go from page 2 to page 3 locally without answering any questions or getting any validation errors, but there `where_did_this_take_place_` is marked as `required` in the YML.

File: crt_portal/cts_forms/views.py [hunk 1]
```diff
@@ -20,11 +20,18 @@ class CRTReportWizard(SessionWizardView):
...
-        form_data = [form.cleaned_data for form in form_list]
-        consolidated_data = dict(ChainMap(*form_data))
-        report_instance = Report(**consolidated_data)
-        report_instance.save()
+        form_data_dict = self.get_all_cleaned_data()
+        m2mfield = form_data_dict.pop('protected_class')
+        r = Report.objects.create(**form_data_dict)
+        r.save()
+
+        for protected in m2mfield:
```
Comment (alexsoble) [inline crt_portal/cts_forms/views.py]: @LindsayYoung I think a code comment would be helpful here to explain why custom handling of the form data is needed.

File: crt_portal/cts_forms/tests.py [hunk 2]
```diff
@@ -1,28 +1,59 @@
...
-        past_question = create_question(description_text='Past Question.', days=-5)
-        url = reverse('crt_forms:detail', args=(past_question.id,))
-        response = self.client.get(url)
-        self.assertContains(response, past_question.description_text)
\ No newline at end of file
+
+from .models import ProtectedClass, Report
+from .forms import WhatHappened, Where, Who, Details, Contact
+
+class Valad_Form_Test(TestCase):
```
Comment (alexsoble) [inline crt_portal/cts_forms/tests.py]: @LindsayYoung This should be "valid," not "valad", same below.

File: crt_portal/cts_forms/tests.py [hunk 3]
```diff
@@ -1,28 +1,59 @@
...
+    """Confirms each form is valid when given valid test data."""
```
Comment (alexsoble) [inline crt_portal/cts_forms/tests.py]: @LindsayYoung Let's add a test case with invalid data to make sure that the form isn't incorrectly marking all input data as valid. I think the current state of the form validation logic is that it's overly permissive — when looking at the form manually I found that the form didn't throw an error when I left required fields blank.

**Comment by LindsayYoung** [comment]
Is it only `where_did_this_take_place` that should be required and isn't, or are there more fields? Please list any other fields that are supposed to be required but aren't.

**Comment by alexsoble** [comment]
Pairing on this with @LindsayYoung, we're noticing that these fields also should be required but aren't required on the form yet: 

+ `who_are_you_reporting_for_`
+ `when_did_this_incident_last_take_place_`

**Comment by alexsoble** [comment]
Hey @LindsayYoung! I went back and checked the form copy for copy-and-paste mistakes like you requested; I don't see any. I also don't see any more questions that need to be changed to "required", beyond the 3 we've identified so far. Once those are fixed and you have a minute to address the comments/questions I left above about testing and spelling, I'll give this my approval and we'll ship! 🚢

File: crt_portal/cts_forms/tests.py (hunk 2 above)
Comment (LindsayYoung) [reply crt_portal/cts_forms/tests.py]: You will find that I am a bad speller, so thanks for finding these things.

**Comment by LindsayYoung** [comment]
Ok, I noticed an error, I merged the `who` and `who_are_you_reporting_for` fields I am going to straighten that out and rename `who_are_you_reporting_for` to `who_are_you_reporting_about` for clarity.

**Comment by LindsayYoung** [comment]
It wasn't the variable names, it was the options. I almost have it straightened out.

**Comment by LindsayYoung** [comment]
@alexsoble I made updates and it is ready for your final review, thanks again!

File: crt_portal/cts_forms/tests.py (hunk 3 above)
Comment (alexsoble) [reply crt_portal/cts_forms/tests.py]: Fixed! Thank you @LindsayYoung

File: crt_portal/cts_forms/tests.py (hunk 2 above)
Comment (alexsoble) [reply crt_portal/cts_forms/tests.py]: 👍

File: crt_portal/cts_forms/views.py (hunk 1 above)
Comment (alexsoble) [reply crt_portal/cts_forms/views.py]: Fixed, thanks @LindsayYoung 😄

**Comment by alexsoble** [comment]
@LindsayYoung When I apply the new migrations and run the app locally, I'm not seeing any checkboxes or form elements under Protected Class, is this the same as what you see? 

protected-class-empty

**Comment by alexsoble** [comment]
Ah, I needed to log in to http://0.0.0.0:8000/report/ and add examples of Protected Classes to the database to test.
protected-class-examples

**Comment by alexsoble** [comment]
In re-checking the form, I noticed that I can still navigate past page 2 without selecting any answers; that suggests to me that validation for the `where_did_this_take_place_` question is still missing. 

However, there's lots of great work on this PR, and I don't want to hold it up until every detail of the form is perfect — we can continue iteratively improving on the form. 

@LindsayYoung I'll approve this and leave it up to you to 🚢if you like, or add more validation if that seems better to you. Cheers!

**Comment by LindsayYoung** [comment]
two test for when, need to add one for where/location

**Comment by LindsayYoung** [comment]
I checked locally and the following are required and have tests: 
    primary complaint, 
    place, 
    violation summary, 
    when, 
    who_reporting_for, 

We can change the requirements as the form changes in the future.

Feel free to take another look @alexsoble and merge when you are done, Thanks!

**Comment by alexsoble** [comment]
@LindsayYoung I tested locally, the place validation is now working! And glad you spotted and fixed that duplicate test.

 🚢 🚢

**Pull request merged**